### ISSUE TITLE
icu build is using foreign_cc despite it being a bazel package

### ISSUE THREAD

**Issue by phlax** [envoyproxy/envoy#26395]
Reading through the original PR that added this #20143 its not clear why icu was added as a foreign_cc package when it has a bazel build

Either way this dependency causes constant network transient flakes as it seems to try and run bazel as a foreign build

**Comment by phlax** [comment]
cc @realtimetodie im wondering why it was built this way and if we can shift this to a proper bazel build - its causing a pretty frequent CI flake, and seems wrong

**Comment by realtimetodie** [comment]
At the time when this extension was written, the implementation of the Bazel build system in libicu was incomplete and did not work.

**Comment by realtimetodie** [comment]
@markusicu can you help?

**Comment by realtimetodie** [comment]
@markusicu can you help us please? As far as I understand, the Bazel build system implementation in the libicu repository is only sufficient to build small parts of its libraries and not the full library.

**Comment by phlax** [comment]
i see

>  We still need a layer outside of Bazel in order to copy the tool output into the source tree, because Bazel on its own does not allow modifying the source tree

this suggests some misunderstanding about how bazel works and how to build hermetically.

Not sure what it means for other bazel projects that want to use the repo.

@realtimetodie if we cant easily resolve upstream build issues then i guess a/nother temporary workaround is as you suggested to strip any other bazel files - but im wondering if that is actually possible given the above - dont some parts _require_ bazel to be built ?

**Comment by realtimetodie** [comment]
No, it seems that Bazel is still optional and only a tiny subset of the libicu libraries are using it.

I will take a look into this issue tonight. I originally wrote the icu.patch file. It removes many BUILD.bazel files from the downloaded source code. However, the latest update of the icu library was unfortunately not done by me, and I think that some new BUILD.bazel files need to be added to the patch.

https://github.com/envoyproxy/envoy/blob/9dc67f4524bac66a0ae926799d41d7a64afe1256/bazel/foreign_cc/icu.patch

**Comment by phlax** [comment]
> was unfortunately not done by me

cc @phlax 

probably we should add some info about updating the dep in repository_locations.bzl

**Comment by phlax** [comment]
> It removes many BUILD.bazel

its probably important to remove the WORKSPACE also i guess - _something_ is triggering bazel in the current build

**Comment by realtimetodie** [comment]
Yes, it is either some WORKSPACE or BUILD file, maybe both. It is unfortunate that the Bazel implementation of libicu is so half-baked. I remember that I tried to make it work under Bazel by adding the missing BUILD definitions, but that quickly grew out of scope.

**Comment by phlax** [comment]
@realtimetodie related to this running a re-compile locally (using RBE) and the icu job is one of the slowest to complete

partly that just may be a function of it using `foreign_cc` but im wondering if we can make this dependency a bit more optimal - my understanding is that we only have to parse country codes

**Comment by realtimetodie** [comment]
Originally, I focused on binary size and not compile time optimization using the [icu feature flags](https://github.com/unicode-org/icu/blob/b6b7b045e9cef2c942efd267bb89c5a545017f0c/icu4c/source/configure#L1449) only. But I think there is still room for optimizations, like turning off conversion or collation using compiler flags. This is mentioned in the [documentation](https://github.com/unicode-org/icu/blob/b6b7b045e9cef2c942efd267bb89c5a545017f0c/docs/userguide/icu4c/build.md#user-configurable-settings). I am working on this in https://github.com/realtimetodie/envoy/tree/icu-config-features.

**Comment by phlax** [comment]
@realtimetodie seems like we are still hitting some bazel issues

```
INFO: Streaming build results to: https://source.cloud.google.com/results/invocations/c179f544-546a-43aa-a11c-33861800d6e7
WARNING: /source/test/proto/BUILD:30:23: in genrule rule //test/proto:bookstore_proto_descriptor: target '//test/proto:bookstore_proto_descriptor' depends on deprecated target '@com_google_protobuf//:well_known_protos': Prefer :well_known_type_protos instead.
INFO: Repository com_github_unicode_org_icu instantiated at:
  /source/WORKSPACE:13:19: in <toplevel>
  /source/bazel/repositories.bzl:260:32: in envoy_dependencies
  /source/bazel/repositories.bzl:468:26: in _com_github_unicode_org_icu
  /source/bazel/repositories.bzl:30:23: in external_http_archive
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_api/bazel/envoy_http_archive.bzl:16:17: in envoy_http_archive
Repository rule http_archive defined at:
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
WARNING: Download from https://github.com/unicode-org/icu/archive/release-72-1.tar.gz failed: class java.io.IOException java.net.SocketTimeoutException: Read timed out
ERROR: An error occurred during the fetch of repository 'com_github_unicode_org_icu':
   Traceback (most recent call last):
	File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_unicode_org_icu/temp10562923677977769683/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
ERROR: /source/WORKSPACE:13:19: fetching http_archive rule //external:com_github_unicode_org_icu: Traceback (most recent call last):
	File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_unicode_org_icu/temp10562923677977769683/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
INFO: Repository kafka_server_binary instantiated at:
  /source/WORKSPACE:13:19: in <toplevel>
```

https://dev.azure.com/cncf/envoy/_build/results?buildId=133388&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=427034ac-ea02-50a9-4444-aee028d5fd94&l=51

trying to understand why this happens so frequently

my thought was that it was due to it being a separate bazel workspace and not respecting our `retries` flag fro some reason, but not sure that is what the error says

~relatedly im trying to track down builds that dont get properly cached and wondering if we see this so often just because it doesnt cache properly for some reason - ill run some tests to try and figure out

**Comment by Atish-iaf** [comment]
> [@realtimetodie](https://github.com/realtimetodie) seems like we are still hitting some bazel issues
> 
> ```
> INFO: Streaming build results to: https://source.cloud.google.com/results/invocations/c179f544-546a-43aa-a11c-33861800d6e7
> WARNING: /source/test/proto/BUILD:30:23: in genrule rule //test/proto:bookstore_proto_descriptor: target '//test/proto:bookstore_proto_descriptor' depends on deprecated target '@com_google_protobuf//:well_known_protos': Prefer :well_known_type_protos instead.
> INFO: Repository com_github_unicode_org_icu instantiated at:
>   /source/WORKSPACE:13:19: in <toplevel>
>   /source/bazel/repositories.bzl:260:32: in envoy_dependencies
>   /source/bazel/repositories.bzl:468:26: in _com_github_unicode_org_icu
>   /source/bazel/repositories.bzl:30:23: in external_http_archive
>   /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_api/bazel/envoy_http_archive.bzl:16:17: in envoy_http_archive
> Repository rule http_archive defined at:
>   /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
> WARNING: Download from https://github.com/unicode-org/icu/archive/release-72-1.tar.gz failed: class java.io.IOException java.net.SocketTimeoutException: Read timed out
> ERROR: An error occurred during the fetch of repository 'com_github_unicode_org_icu':
>    Traceback (most recent call last):
> 	File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
> 		download_info = ctx.download_and_extract(
> Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_unicode_org_icu/temp10562923677977769683/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
> ERROR: /source/WORKSPACE:13:19: fetching http_archive rule //external:com_github_unicode_org_icu: Traceback (most recent call last):
> 	File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
> 		download_info = ctx.download_and_extract(
> Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_unicode_org_icu/temp10562923677977769683/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
> INFO: Repository kafka_server_binary instantiated at:
>   /source/WORKSPACE:13:19: in <toplevel>
> ```
> 
> https://dev.azure.com/cncf/envoy/_build/results?buildId=133388&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=427034ac-ea02-50a9-4444-aee028d5fd94&l=51
> 
> trying to understand why this happens so frequently
> 
> my thought was that it was due to it being a separate bazel workspace and not respecting our `retries` flag fro some reason, but not sure that is what the error says
> 
> ~relatedly im trying to track down builds that dont get properly cached and wondering if we see this so often just because it doesnt cache properly for some reason - ill run some tests to try and figure out

Hi @phlax 
I am facing this issue consistently while building envoy v1.33.0 in a container. Is this fixed ?

**Comment by phlax** [comment]
sadly not - its a perennial issue - we mostly avoid it in our ci by using a client cache so it doesnt get downloaded as often - but its as unreliable as ever

the foreign_cc thing i think is unrelated - and iirc the bazel setup doesnt work (at least for us) - but not sure about that

**Comment by markusicu** [comment]
> As far as I understand, the Bazel build system implementation in the libicu repository is only sufficient to build small parts of its libraries and not the full library.

Correct. We added the Bazel files only for use by ICU team members for updating ICU to new versions of Unicode. https://unicode-org.github.io/icu/processes/unicode-update.html

**Comment by realtimetodie** [comment]
The [Language](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/language_filter) HTTP filter only uses the following imports from unicode [ICU4C](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/):

- unicode/locid.h
- unicode/utypes.h
- [unicode/localematcher.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html)

TLDR; as far as I remember the [unicode/localematcher.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class is also used by the Chromium web browser project for the V8 [Intl API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) bindings.

Unfortunately you need to compile a lot of unnecessary stuff just to get to the [unicode/localematcher.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class. Alternatively, I could provide a very big patch (that could potentially break with every ICU4C release) that would modularize the ICU4C Bazel build setup. Or, the ICU4C team would modularize the [unicode/localematcher.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class.

Personally I replaced the Language HTTP filter using parts from the Rust implementation [icu4x|https://github.com/unicode-org/icu4x] a long time ago.

@markusicu Do you want to provide a solution to modularize the compilation of the ICU4C with Bazel?

**Comment by realtimetodie** [comment]
We can also remove the [Language](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/language_filter) HTTP filter if building it causes so much trouble, I don't care.
